README.md
# Timewrap

Timewrap is a "Time-Wrapper" through which you can access different
time-sources, Elixir and Erlang offers you. Other than that you 
can implement on your own.

Also, _Timewrap_ can do the time-warp, freeze, and unfreeze a 
`Timewrap.Timer`.

You can instantiate different `Timewrap.Timer`s, registered and
supervised by `:name`.

The `Timewrap.TimeSupervisor` is started with the `Timewrap.Application`
and implicitly starts the default timer `:default_timer`. This
one is used whenever you call Timewrap-functions without a 
timer given as the first argument.

### Examples:

    use Timewrap # imports some handy Timewrap-functions.

#### With default Timer

    Timewrap.freeze_timer()
    item1 = %{ time: current_time() }
    :timer.sleep(1000)
    item2 = %{ time: current_time() }
    assert item1.time == item2.time


#### Transactions with a given and frozen time

    with_frozen_timer(~N[1964-08-31 06:00:00Z], fn ->
      ... do something while `current_time` will 
      always return the given timestamp within this
      block...
    end )

#### Start several independent timers

    {:ok, today} = new_timer(:today)
    {:ok, next_week} = new_timer(:next_week)
    freeze_time(:today, ~N[2019-02-11 09:00:00])
    freeze_time(:next_week, ~N[2019-02-18 09:00:00])
    ... do something ...
    unfreeze_time(:today)
    unfreeze_time(:next_week)


## Installation

From [available Hex package](https://hex.pm/docs/publish), the package can be installed
by adding `timewrap` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:timewrap, "~> 0.1.0"}
  ]
end
```
  
### Configuration

  `config/config.exs`

      config :timewrap,
        timer: :default,
        unit: :second,
        calendar: Calendar.ISO,
        representation: :unix

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
be found at [https://hexdocs.pm/timewrap](https://hexdocs.pm/timewrap).



CHANGELOG.md
# CHANGELOG

## Initial commit





test/timewrap_test.exs
defmodule TimewrapTest do
  use ExUnit.Case
  doctest Timewrap

  use Timewrap

  setup _ do
    {:ok, timer} = Timewrap.new_timer(:test_timer)

    on_exit(fn -> Timewrap.release_timer(timer) end)

    {:ok, %{t: timer}}
  end

  describe "Timewrap in tests" do
    test "tests start with RealTime", %{t: timer} do
      assert is_pid(timer)
      assert [mode: :transparent, unit: :second] == Timer.mode(timer)
    end
  end
end


mix.exs
defmodule Timewrap.MixProject do
  use Mix.Project

  def project do
    [
      app: :timewrap,
      version: "0.1.0",
      elixir: "~> 1.8",
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description: description(),
      package: package(),
      # The main page in the docs
      docs: [
        main: "readme",
        logo: "assets/images/logo.png",
        extras: ["README.md", "LICENSE.md", "CHANGELOG.md"]
      ]
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger],
      mod: {Timewrap.Application, []}
    ]
  end

  def description,
    do: ~S"""
    **Timewrap** is a simple _wrapper_ you can use to access
    time. Other than that, you can freeze and unfreeze different
    timers. Very handy in tests.
    """

  def package() do
    [
      name: "timewrap",
      files: ["lib", "mix.exs", "README*", "LICENSE*", "CHANGELOG*"],
      maintainers: ["andreas@altendorfer.at"],
      licenses: ["MIT"],
      links: %{
        "GitHub" => "https://github.com/iboard/timewrap",
        "Documentation" => "https://hexdocs.pm/timewrap/readme.html"
      }
    ]
  end

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      {:ex_doc, "~> 0.16", only: :dev, runtime: false}
    ]
  end
end


config/config.exs
# This file is responsible for configuring your application
# and its dependencies with the aid of the Mix.Config module.
use Mix.Config

# This configuration is loaded before any dependency and is restricted
# to this project. If another project depends on this project, this
# file won't be loaded nor affect the parent project. For this reason,
# if you want to provide default values for your application for
# third-party users, it should be done in your "mix.exs" file.

# You can configure your application as:
#
#     config :timewrap, key: :value
#
# and access this configuration in your application as:
#
#     Application.get_env(:timewrap, :key)
#
# You can also configure a third-party app:
#
#     config :logger, level: :info
#

# It is also possible to import configuration files, relative to this
# directory. For example, you can emulate configuration per environment
# by uncommenting the line below and defining dev.exs, test.exs and such.
# Configuration from the imported file will override the ones defined
# here (which is why it is important to import them last).
#

config :timewrap,
  timer: :default,
  unit: :second,
  calendar: Calendar.ISO,
  representation: :unix

#     import_config "#{Mix.env()}.exs"


lib/timewrap/timer.ex
defmodule Timewrap.Timer do
  require Logger
  use Agent

  alias __MODULE__

  def start_link(opts) do
    module =
      case Keyword.fetch(opts, :name) do
        {:ok, name} -> name
        :error -> :default_timer
      end

    Agent.start_link(
      fn -> [mode: :transparent, unit: :second] end,
      name: module
    )
  end

  def current_time(timer \\ :default_timer) do
    case mode(timer) do
      [mode: :transparent, unit: unit] -> System.system_time(unit)
      [mode: :frozen, unit: _, time: time] -> time
      rc -> Logger.error("Unsupported options:" <> inspect(rc, pretty: true))
    end
  end

  def freeze(timer \\ :default_timer, time \\ nil) do
    Agent.update(timer, fn state ->
      unit = Keyword.get(state, :unit)
      [mode: :frozen, unit: unit, time: time || System.system_time(unit)]
    end)

    Agent.get(timer, fn s -> Keyword.get(s, :time) end)
  end

  def unfreeze(timer \\ :default_timer) do
    Agent.update(timer, fn state ->
      [mode: :transparent, unit: Keyword.get(state, :unit)]
    end)

    current_time(timer)
  end

  def mode(timer \\ :default_timer) do
    Agent.get(timer, fn state -> state end)
  end
end


lib/timewrap.ex
defmodule Timewrap do
  @moduledoc """
  Timewrap is a "Time-Wrapper" through which you can access different
  time-sources, Elixir and Erlang offers you. Other than that you 
  can implement on your own.

  Also, _Timewrap_ can do the time-warp, freeze, and unfreeze a 
  `Timewrap.Timer`.

  You can instantiate different `Timewrap.Timer`s, registered and
  supervised by `:name`.

  The `Timewrap.TimeSupervisor` is started with the `Timewrap.Application`
  and implicitly starts the default timer `:default_timer`. This
  one is used whenever you call Timewrap-functions without a 
  timer given as the first argument.

  ### Configuration

  `config/config.exs`

      config :timewrap,
        timer: :default,
        unit: :second,
        calendar: Calendar.ISO,
        representation: :unix


  ### Examples:

      use Timewrap # imports some handy Timewrap-functions.

  #### With default Timer

      Timewrap.freeze_timer()
      item1 = %{ time: current_time() }
      :timer.sleep(1000)
      item2 = %{ time: current_time() }
      assert item1.time == item2.time


  #### Transactions with a given and frozen time

      with_frozen_timer(~N[1964-08-31 06:00:00Z], fn ->
        ... do something while `current_time` will 
        always return the given timestamp within this
        block...
      end )

  #### Start several independent timers

      {:ok, today} = new_timer(:today)
      {:ok, next_week} = new_timer(:next_week)
      freeze_time(:today, ~N[2019-02-11 09:00:00])
      freeze_time(:next_week, ~N[2019-02-18 09:00:00])
      ... do something ...
      unfreeze_time(:today)
      unfreeze_time(:next_week)

  """

  defmacro __using__(_args) do
    quote do
      alias Timewrap.Timer

      import Timewrap,
        only: [
          current_time: 0,
          freeze_time: 0,
          freeze_time: 1,
          unfreeze_time: 0,
          unfreeze_time: 1,
          new_timer: 1,
          release_timer: 1
        ]
    end
  end

  @doc """
  Get the `current_time` in the format you've configured.

  ### Configuration

  TODO: Describe configuration here.

  ### Examples

      iex> Timewrap.current_time() == System.system_time(:second)
      true

  """
  def current_time(timer \\ :default_timer) do
    Timewrap.Timer.current_time(timer)
  end

  @doc """
  Freeze the current time. All calls to `current_time` will return
  the same value until you call `unfreeze`.

  ### Example:

      iex> frozen = Timewrap.freeze_time()
      iex> :timer.sleep(1001)
      iex> assert frozen == Timewrap.current_time()
      true

  """
  def freeze_time(timer \\ :default_timer, time \\ nil) do
    freeze(timer, time)
  end

  @doc """
  Unfreeze a frozen time. If the Timer is not frozen, this function
  is a _noop_.

  ### Example:

      iex> frozen = Timewrap.freeze_time()
      iex> :timer.sleep(1001)
      iex> assert frozen == Timewrap.current_time()
      iex> reseted = Timewrap.unfreeze_time()
      iex> assert reseted == System.system_time(:second)
      iex> assert reseted != frozen
      true

  """
  def unfreeze_time(timer \\ :default_timer) do
    unfreeze(timer)
  end

  @doc """
  Start a new timer-agent. A supervised worker of
  `Timewrap.TimeSupervisor`.

  ### Example:

      iex> use Timewrap
      iex> {:ok, t} = new_timer(:mytime)
      iex> assert is_pid(t)
      iex> :timer.sleep(100)
      iex> Timewrap.release_timer(t)
      :ok

  """
  def new_timer(name) do
    {:ok, pid} = Timewrap.TimeSupervisor.start_timer(name)
    {:ok, pid}
  end

  @doc """
  Release a running timer terminates the process and removes
  it from supervision.

  ### Example:

      iex> use Timewrap
      iex> {:ok, t} = new_timer(:mytime)
      iex> :timer.sleep(100)
      iex> release_timer(t)
      iex> Process.alive?(t)
      false

  """
  def release_timer(pid) do
    if Process.alive?(pid) do
      Timewrap.TimeSupervisor.stop_timer(pid)
    end
  end

  @doc """
  Execute a given block with a timer frozen at the given time

  ### Example

       iex> use Timewrap
       iex> "1964-08-31 06:00:00+0100" 
       iex> |> Timewrap.with_frozen_time(fn() ->
       iex>   assert current_time() == -168372000
       iex> end) 
       true
  """
  def with_frozen_time(time, fun) do
    {:ok, dt, offset} = DateTime.from_iso8601(time)

    freeze_time(:default_timer, DateTime.to_unix(dt) + offset)
    rc = fun.()
    unfreeze_time(:default_timer)
    rc
  end

  defp freeze(timer, time), do: Timewrap.Timer.freeze(timer, time)

  defp unfreeze(timer), do: Timewrap.Timer.unfreeze(timer)
end
